src/Tooltip/Tooltip.js

```diff
@@ -4,91 +4,183 @@ import ReactDOM from 'react-dom';
+        return 'start';
       }
-    } else if (position === 'top' || position === 'bottom') {
-      if (renderedPosition.right > viewportWidth) {
-        anchor = 'end';
-      } else if (renderedPosition.left < 0) {
-        anchor = 'start';
+
+      if (clearanceEnd < tooltipOverflow) {
+        return 'end';
       }
     }
 
     return anchor;
   }
 
-  getPosition(position, renderedPosition, viewportWidth, viewportHeight) {
+  getAnchor(isVertical, anchor, clearance, tooltipWidth, tooltipHeight,
+    triggerRect) {
+    // Calculate the ideal anchor.
+    if (isVertical) {
+      return this.transformAnchor(anchor, clearance.left, clearance.right,
+        tooltipWidth, triggerRect.width);
+    }
+
+    return this.transformAnchor(anchor, clearance.top, clearance.bottom,
+      tooltipHeight, triggerRect.height);
+  }
+
+  getCoordinates(triggerRect, position, clearance, tooltipWidth,
+    tooltipHeight) {
+    // Calculate the coordinates of the tooltip content.
+    if (position === 'top') {
+      return {
+        left: triggerRect.left + triggerRect.width / 2,
+        top: triggerRect.top - tooltipHeight + ARROW_SIZE
+      };
+    } else if (position === 'right') {
+      return {
+        left: triggerRect.right,
+        top: triggerRect.top + triggerRect.height / 2
+      };
+    } else if (position === 'bottom') {
+      return {
+        left: triggerRect.left + triggerRect.width / 2,
+        top: triggerRect.bottom
+      };
+    }
+
+    return {
+      left: triggerRect.left - tooltipWidth + ARROW_SIZE,
+      top: triggerRect.top + triggerRect.height / 2
+    };
+  }
+
+  isVertical(position) {
+    if (position === 'left' || position === 'right') {
+      return false;
+    }
+
+    return true;
+  }
+
+  getPosition(position, clearance, tooltipWidth, tooltipHeight) {
     // Change the position if the tooltip will be rendered off the screen.
-    if (position === 'right' && renderedPosition.right > viewportWidth) {
-      position = 'left';
-    } else if (position === 'left' && renderedPosition.left < 0) {
+    if (position === 'left' && clearance.left < tooltipWidth) {
       position = 'right';
-    } else if (position === 'top' && renderedPosition.top < 0) {
+    } else if (position === 'right' && clearance.right < tooltipWidth) {
+      position = 'left';
+    }
+
+    if (position === 'top' && clearance.top < tooltipHeight) {
       position = 'bottom';
-    } else if (position === 'bottom' &&
-      renderedPosition.bottom > viewportHeight) {
+    } else if (position === 'bottom' && clearance.bottom < tooltipHeight) {
       position = 'top';
     }
 
     return position;
   }
 
-  getIdealLocation() {
-    let anchor = this.state.anchor || this.props.anchor;
-    let position = this.state.position || this.props.position;
+  getIdealLocation(anchor, position) {
+    let isVertical = this.isVertical(position);
+    let triggerNode = ReactDOM.findDOMNode(this.refs.triggerNode);
```

Comment:
i believe `this.refs.triggerNode` will already give you the dom node in this version of react